allennlp/experiments/drivers/train_driver.py

```diff
@@ -0,0 +1,182 @@
+                 serialization_prefix: Optional[str] = None,
+                 cuda_device: int = -1) -> None:
+        """
+        Parameters
+        ----------
+        model : ``Model``, required.
+            An AllenNLP model to be optimized. Pytorch Modules can also be optimized if
+            their ``forward`` method returns a dictionary with a "loss" key, containing a
+            scalar tensor representing the loss function to be optimized.
+        optimizer : ``torch.nn.Optimizer``, required.
+            An instance of a Pytorch Optimizer, instantiated with the parameters of the
+            model to be optimized.
+        iterator : ``DataIterator``, required.
+            A method for iterating over a ``Dataset``, yielding padded indexed batches.
+        train_dataset : ``Dataset``, required.
+            A ``Dataset`` to train on.
+        validation_dataset : ``Dataset``, optional, (default = None).
+            A ``Dataset`` to evaluate on.
+        patience : int, optional (default=2)
+            Number of epochs to be patient before early stopping.  I.e., if the ``validation_metric``
+            does not improve for this many epochs, we will stop training.
+        batch_size : int, optional (default = 32)
+            Batch size to use when training.
+        num_epochs : int, optional (default = 20)
+            Number of training epochs.
+        serialization_prefix : str, optional (default=None)
+            Path to directory for saving and loading model files. Models will not be saved if
+            this parameter is not passed.
+        cuda_device : int, optional (default = -1)
+            An integer specifying the CUDA device to use. If -1, the CPU is used.
+            Multi-gpu training is not currently supported, but will be once the
+            Pytorch DataParallel API stabilises.
+        """
+        self._model = model
+        self._iterator = iterator
+        self._optimizer = optimizer
+        self._train_dataset = train_dataset
+        self._validation_dataset = validation_dataset
+
+        self._patience = patience
+        self._batch_size = batch_size
+        self._num_epochs = num_epochs
+        self._serialization_prefix = serialization_prefix
+        self._cuda_device = cuda_device
+
+        if self._cuda_device >= 0:
+            self._model = self._model.cuda(self._cuda_device)
+
+    @overrides
+    def run(self) -> None:
+        epoch_counter = 0
+        # Resume from serialization path if it contains a saved model.
+        if self._serialization_prefix is not None:
+            if any(["model_state_epoch_" in x
+                    for x in os.listdir(self._serialization_prefix)]):
+                epoch_counter = self._restore_checkpoint()
+
+        for epoch in range(epoch_counter, self._num_epochs):
+            train_loss = 0.0
+            val_loss = 0.0
+
+            # Set the model to "train" mode.
+            self._model.train()
```

Comment:
Yeah, that's an incredibly confusing naming choice from pytorch...  Your comment is definitely necessary.